In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

In [350]:
songs_dataset = pd.read_json('MasterSongList.json')

In [3]:
songs_dataset.shape

(36733, 16)

In [4]:
from string import punctuation
from sklearn.feature_extraction.stop_words import ENGLISH_STOP_WORDS
from nltk.stem.snowball import SnowballStemmer

translator = str.maketrans('', '', punctuation)
stemmer = SnowballStemmer('english')

In [5]:
def clean_text(raw_text):
#   initialize empty clean_words list
    clean_words = []
    raw_text = raw_text.translate(translator)
    raw_text_split = raw_text.split()

    for word in raw_text_split:
        if word not in ENGLISH_STOP_WORDS:
            clean_words.append(stemmer.stem(word))
    return ' '.join(clean_words)

In [6]:
songs_have_lyrics = songs_dataset[songs_dataset['lyrics_features'].str.len() != 0]

In [7]:
#dropping the rows in which column['lyrics_features'] is an empty list — converts to string though...
#songs_dataset_has_lyrics = songs_dataset[songs_dataset['lyrics_features'].str.len() != 0]

In [8]:
#songs_dataset_has_lyrics = songs_dataset_has_lyrics.reset_index

In [9]:
songs_have_lyrics.shape

(20931, 16)

In [10]:
songs_have_lyrics.reset_index()

index                                   _id  \
0          0  {'$oid': '52fdfb440b9398049f3d7a8c'}   
1          1  {'$oid': '52fdfb3d0b9398049f3cbc8e'}   
2          2  {'$oid': '52fdfb420b9398049f3d3ea5'}   
3          3  {'$oid': '52fdfb410b9398049f3d1eac'}   
4          4  {'$oid': '52fdfb400b9398049f3d0b19'}   
5          5  {'$oid': '52fdfb3e0b9398049f3cdd7f'}   
6          6  {'$oid': '52fdfb3f0b9398049f3ce0f0'}   
7          7  {'$oid': '52fdfb440b9398049f3d7a7d'}   
8          8  {'$oid': '52fdfb400b9398049f3d0d47'}   
9          9  {'$oid': '52fdfb400b9398049f3d0dc7'}   
10        10  {'$oid': '52fdfb430b9398049f3d52a7'}   
11        11  {'$oid': '52fdfb400b9398049f3d102c'}   
12        12  {'$oid': '52fdfb410b9398049f3d1908'}   
13        13  {'$oid': '52fdfb400b9398049f3cf917'}   
14        14  {'$oid': '52fdfb400b9398049f3d0b1e'}   
15        15  {'$oid': '52fdfb400b9398049f3d0db1'}   
16        16  {'$oid': '52fdfb3d0b9398049f3cbcb7'}   
17        17  {'$oid': '52fdfb3f0b9398049f3ce0b3'}   
18        18  {'$oid': '52fdfb3e0b9398049f3cd654'}   
19        19  {'$oid': '52fdfb410b9398049f3d1546'}   
20        20  {'$oid': '52fdfb410b9398049f3d1ec5'}   
21        21  {'$oid': '52fdfb400b9398049f3d1031'}   
22        23  {'$oid': '52fdfb400b9398049f3d0db7'}   
23        24  {'$oid': '52fdfb410b9398049f3d193a'}   
24        25  {'$oid': '52fdfb3d0b9398049f3cc6f6'}   
25        26  {'$oid': '52fdfb3d0b9398049f3cbc9d'}   
26        27  {'$oid': '52fdfb400b9398049f3d0db8'}   
27        28  {'$oid': '52fdfb400b9398049f3d0b11'}   
28        29  {'$oid': '52fdfb410b9398049f3d1923'}   
29        30  {'$oid': '52fdfb3e0b9398049f3ccb45'}   
...      ...                                   ...   
20901  36544  {'$oid': '52fdfb3d0b9398049f3cc0cb'}   
20902  36545  {'$oid': '52fdfb3e0b9398049f3cc94c'}   
20903  36560  {'$oid': '52fdfb400b9398049f3d0f41'}   
20904  36573  {'$oid': '52fdfb410b9398049f3d260c'}   
20905  36578  {'$oid': '52fdfb470b9398049f3dbb2d'}   
20906  36584  {'$oid': '52fdfb3d0b9398049f3cc2de'}   
20907  36602  {'$oid': '52fdfb400b9398049f3d04c8'}   
20908  36603  {'$oid': '52fdfb440b9398049f3d6eed'}   
20909  36604  {'$oid': '52fdfb3e0b9398049f3cc99c'}   
20910  36606  {'$oid': '52fdfb400b9398049f3d0604'}   
20911  36608  {'$oid': '52fdfb450b9398049f3d7b52'}   
20912  36614  {'$oid': '52fdfb450b9398049f3d8817'}   
20913  36620  {'$oid': '52fdfb3f0b9398049f3ce2f4'}   
20914  36623  {'$oid': '52fdfb450b9398049f3d91c8'}   
20915  36624  {'$oid': '52fdfb3d0b9398049f3cbf2b'}   
20916  36625  {'$oid': '52fdfb3e0b9398049f3cdc90'}   
20917  36635  {'$oid': '52fdfb400b9398049f3d1138'}   
20918  36637  {'$oid': '52fdfb440b9398049f3d6107'}   
20919  36649  {'$oid': '52fdfb3f0b9398049f3ceeb4'}   
20920  36654  {'$oid': '52fdfb3d0b9398049f3cbf41'}   
20921  36663  {'$oid': '52fdfb400b9398049f3cfd90'}   
20922  36667  {'$oid': '52fdfb420b9398049f3d45ce'}   
20923  36676  {'$oid': '52fdfb400b9398049f3d0c34'}   
20924  36681  {'$oid': '52fdfb460b9398049f3d9c0f'}   
20925  36688  {'$oid': '52fdfb440b9398049f3d6f46'}   
20926  36702  {'$oid': '52fdfb3f0b9398049f3cecd8'}   
20927  36716  {'$oid': '52fdfb420b9398049f3d3a51'}   
20928  36720  {'$oid': '52fdfb430b9398049f3d57a8'}   
20929  36723  {'$oid': '52fdfb460b9398049f3da36b'}   
20930  36725  {'$oid': '52fdfb440b9398049f3d6895'}   

                                                   album  \
0                                  Gangnam Style (강남스타일)   
1                                                 Native   
2                                      Party Rock Anthem   
3                                              Gentleman   
4                                           On The Floor   
5                                          The Lazy Song   
6                                    Rolling In The Deep   
7                                          Call Me Maybe   
8            Somebody That I Used To Know (Feat. Kimbra)   
9                               What Makes You Beautiful  

In [11]:
lyricsFeatureList = []
for lyrics_feature in songs_have_lyrics['lyrics_features']:
    lyricsFeatureList.append(' '.join(lyrics_feature))

In [ ]:
len(lyricsFeatureList)

In [63]:
clean_lyricsFeatureList = []
for song in lyricsFeatureList:
    clean_lyricsFeatureList.append(clean_text(song))
    #print('clean_lyricsFeatureList is getting: ', clean_text(song))

In [79]:
sd = songs_have_lyrics.copy()

se = pd.Series(clean_lyricsFeatureList)
sd['cleaned_lyrics'] = se.values

In [81]:
sd.to_json('songs_have_lyrics.json')

In [382]:
songs_df = pd.read_json('songs_have_lyrics.json')

In [233]:
songs_df

level_0  index                                   _id  \
0            0      0  {'$oid': '52fdfb440b9398049f3d7a8c'}   
1            1      1  {'$oid': '52fdfb3d0b9398049f3cbc8e'}   
2            2     10  {'$oid': '52fdfb430b9398049f3d52a7'}   
3            3    100  {'$oid': '52fdfb3e0b9398049f3ccb47'}   
4            4   1000  {'$oid': '52fdfb400b9398049f3d0bf1'}   
5            5  10001  {'$oid': '52fdfb3f0b9398049f3cf4b4'}   
6            6  10002  {'$oid': '52fdfb420b9398049f3d317c'}   
7            7  10004  {'$oid': '52fdfb440b9398049f3d62a8'}   
8            8  10005  {'$oid': '52fdfb420b9398049f3d4221'}   
9            9  10006  {'$oid': '52fdfb460b9398049f3d94f5'}   
10          10  10007  {'$oid': '52fdfb3e0b9398049f3ccc73'}   
11          11  10008  {'$oid': '52fdfb440b9398049f3d705c'}   
12          12   1001  {'$oid': '52fdfb3e0b9398049f3cc9af'}   
13          13  10011  {'$oid': '52fdfb460b9398049f3daac9'}   
14          14  10012  {'$oid': '52fdfb3d0b9398049f3cc412'}   
15          15  10014  {'$oid': '52fdfb440b9398049f3d6d18'}   
16          16  10015  {'$oid': '52fdfb440b9398049f3d62a6'}   
17          17  10016  {'$oid': '52fdfb430b9398049f3d4985'}   
18          18  10018  {'$oid': '52fdfb440b9398049f3d6465'}   
19          19  10020  {'$oid': '52fdfb400b9398049f3cfb3a'}   
20          20  10021  {'$oid': '52fdfb460b9398049f3dae55'}   
21          21  10022  {'$oid': '52fdfb400b9398049f3d010a'}   
22          22  10023  {'$oid': '52fdfb470b9398049f3dbba6'}   
23          23  10024  {'$oid': '52fdfb3f0b9398049f3cf2b4'}   
24          24  10025  {'$oid': '52fdfb3f0b9398049f3ce319'}   
25          25  10026  {'$oid': '52fdfb3f0b9398049f3ceea4'}   
26          26  10027  {'$oid': '52fdfb3f0b9398049f3cf170'}   
27          27  10028  {'$oid': '52fdfb3f0b9398049f3ce597'}   
28          28  10029  {'$oid': '52fdfb3e0b9398049f3ccd83'}   
29          29   1003  {'$oid': '52fdfb460b9398049f3d94dc'}   
...        ...    ...                                   ...   
20901    20901   9961  {'$oid': '52fdfb400b9398049f3d09e4'}   
20902    20902   9962  {'$oid': '52fdfb440b9398049f3d6464'}   
20903    20903   9963  {'$oid': '52fdfb410b9398049f3d1d9a'}   
20904    20904   9964  {'$oid': '52fdfb3e0b9398049f3ccd5e'}   
20905    20905   9965  {'$oid': '52fdfb3e0b9398049f3cd168'}   
20906    20906   9966  {'$oid': '52fdfb3f0b9398049f3ce751'}   
20907    20907   9967  {'$oid': '52fdfb430b9398049f3d5846'}   
20908    20908   9968  {'$oid': '52fdfb440b9398049f3d6bda'}   
20909    20909   9969  {'$oid': '52fdfb460b9398049f3d98f1'}   
20910    20910    997  {'$oid': '52fdfb3e0b9398049f3cd4f2'}   
20911    20911   9971  {'$oid': '52fdfb420b9398049f3d4007'}   
20912    20912   9972  {'$oid': '52fdfb430b9398049f3d596d'}   
20913    20913   9975  {'$oid': '52fdfb3d0b9398049f3cc37c'}   
20914    20914   9976  {'$oid': '52fdfb470b9398049f3db9f9'}   
20915    20915   9977  {'$oid': '52fdfb460b9398049f3d9fc3'}   
20916    20916   9978  {'$oid': '52fdfb420b9398049f3d3d3b'}   
20917    20917    998  {'$oid': '52fdfb450b9398049f3d8a07'}   
20918    20918   9980  {'$oid': '52fdfb3f0b9398049f3ce9e5'}   
20919    20919   9981  {'$oid': '52fdfb470b9398049f3db3c2'}   
20920    20920   9982  {'$oid': '52fdfb460b9398049f3d937f'}   
20921    20921   9983  {'$oid': '52fdfb470b9398049f3dbf96'}   
20922    20922   9984  {'$oid': '52fdfb420b9398049f3d2efa'}   
20923    20923   9985  {'$oid': '52fdfb400b9398049f3d0c63'}   
20924    20924   9989  {'$oid': '52fdfb420b9398049f3d2f28'}   
20925    20925    999  {'$oid': '52fdfb3e0b9398049f3cc9a4'}   
20926    20926   9991  {'$oid': '52fdfb460b9398049f3da5fb'}   
20927    20927   9992  {'$oid': '52fdfb460b9398049f3da2b8'}   
20928    20928   9995  {'$oid': '52fdfb3f0b9398049f3cf346'}   
20929    20929   9996  {'$oid': '52fdfb460b9398049f3d98f4'}   
20930    20930   9997  {'$oid': '52fdfb400b9398049f3d0627'}   

                                             album                     artist  \
0                  

In [160]:
songs_df.shape

(20931, 14)

In [65]:
#songs_dataset_has_lyrics.columns

In [66]:
#for item in songs_dataset_has_lyrics["lyrics_features"]:
 #   songs_dataset_has_lyrics.replace(to_replace=item, value=(lyricsFeatureList[i] for i in range(len(lyricsFeatureList))), inplace=True)

# Bag of Words

In [117]:
count_vectorizer = CountVectorizer()
bag_of_words = count_vectorizer.fit_transform(sd['cleaned_lyrics'])

In [275]:
print(bag_of_words)

  (0, 53625)	1
  (0, 33952)	1
  (0, 32375)	2
  (0, 41017)	2
  (0, 4912)	4
  (0, 66721)	2
  (0, 25128)	2
  (0, 42683)	6
  (0, 62810)	2
  (0, 63428)	1
  (0, 53465)	1
  (0, 25171)	1
  (0, 39177)	1
  (0, 65993)	1
  (0, 17762)	1
  (0, 62746)	1
  (0, 31905)	1
  (0, 24479)	1
  (0, 67575)	1
  (0, 42618)	1
  (0, 66369)	1
  (0, 24662)	1
  (0, 49017)	1
  (0, 38863)	1
  (0, 40712)	1
  :	:
  (20930, 9659)	1
  (20930, 16985)	1
  (20930, 61730)	1
  (20930, 25711)	1
  (20930, 45675)	2
  (20930, 28753)	2
  (20930, 27162)	1
  (20930, 42291)	1
  (20930, 6117)	1
  (20930, 55471)	1
  (20930, 52497)	2
  (20930, 47209)	1
  (20930, 42183)	1
  (20930, 66397)	1
  (20930, 25983)	1
  (20930, 58205)	1
  (20930, 29072)	1
  (20930, 28937)	1
  (20930, 62976)	2
  (20930, 68092)	1
  (20930, 35638)	2
  (20930, 35887)	1
  (20930, 52985)	1
  (20930, 48023)	2
  (20930, 60932)	1


In [276]:
simplified_moods = []
for mood in sd['moods']:
    if 'happy' in mood:
        simplified_moods.append('happy')
    elif 'sad' in mood:
        simplified_moods.append('sad')
    else: 
        simplified_moods.append(mood)
len(simplified_moods)

20931

In [279]:
se2 = pd.Series(simplified_moods)
sd['simplified_moods'] = se2.values

### Picking all the 'happy' & 'sad' songs and Balancing the Data

In [280]:
sd[(sd['simplified_moods'] == 'happy') | (sd['simplified_moods'] == 'sad')]

artist  \
1                       OneRepublic   
2                             LMFAO   
3                               PSY   
5                        Bruno Mars   
9                     One Direction   
12                       Katy Perry   
15                       Bruno Mars   
16                      Miley Cyrus   
19                      Nicki Minaj   
20                        will.i.am   
23     Hannah Montana (Miley Cyrus)   
24                          Rihanna   
26                             Pink   
27                         Jessie J   
29                          Rihanna   
30                          Beyoncé   
32                  Imagine Dragons   
34                     Selena Gomez   
35                    One Direction   
40                      Demi Lovato   
41          Macklemore & Ryan Lewis   
42                       Bruno Mars   
43                    Calvin Harris   
44                  Christina Perri   
46                     Taylor Swift   
49                          Rihanna   
56                          Pitbull   
57                  Imagine Dragons   
58                     Taylor Swift   
63                  The Cranberries   
...                             ...   
34925                    Snailhouse   
34941                 Janet Jackson   
34991               Musiq Soulchild   
35007                  Mia Doi Todd   
35022              Rosemary Clooney   
35032                      Jay Park   
35047                     Neko Case   
35052                   Loney, Dear   
35209              The Darling Buds   
35249                  Simon Joyner   
35289               Brooke Waggoner   
35400                  David Ruffin   
35468                    Arab Strap   
35538                    Cal Tjader   
35583                    Perry Como   
35619              Assembly Of Dust   
35716                    Wooden Sky   
35924                   Rita Moreno   
35936             Adalberto Álvarez   
36017                 Grateful Dead   
36067              Assembly Of Dust   
36095           Great Lake Swimmers   
36164                Spandau Ballet   
36239                    Tracy Byrd   
36306                Marissa Nadler   
36452                  Akron/Family   
36523                Gerry Mulligan   
36537                 The Good Life   
36604                George Michael   
36667                Richard Hawley   

                                                 context  \
1                                            [energetic]   
2           [energetic, energetic, energetic, energetic]   
3             [party, party, party, party, party, party]   
5                                         [just woke up]   
9      [studying / working, studying / working, study...   
12                                 [party, party, party]   
15     [studying / working, studying / working, study...   
16                                           [energetic]   
19                                               [party]   
20            [party, party, party, party, party, party]   
23     [studying / working, studying / working, study...   
24                                 [party, party, party]   
26                                           [energetic]   
27     [studying / working, studying / working, study...   
29                                 [party, party, party]   
30                                               [party]   
32                                           [energetic]   
34                                           [energetic]   
35                                           [energetic]   
40                                           [energetic]   
41                                           [energetic]   
42                                           [energetic]   
43                                           [energetic]   
44                                           [energetic]   
46          [energetic, energetic, energetic, energetic]   
49                                               [

In [281]:
sd[(sd['simplified_moods'] == 'happy')].shape

(1757, 10)

In [282]:
#sampling 1249 'happy':
sd_happy = sd[(sd['simplified_moods'] == 'happy')].sample(n=1249, replace=True)

In [283]:
sd_happy.shape

(1249, 10)

In [284]:
sd[(sd['simplified_moods'] == 'sad')].shape

(1249, 10)

***Concatenating the songs with lyrics in either 'happy' or 'sad' mood***

In [285]:
happy_or_sad_DF = pd.concat((sd_happy, sd[(sd['simplified_moods'] == 'sad')]), axis=0)

In [286]:
happy_or_sad_DF.shape

(2498, 10)

# Bag of Words (BOW)

In [333]:
bag_of_words_happy_sad = count_vectorizer.fit_transform(happy_or_sad_DF['cleaned_lyrics'])

In [334]:
bag_of_words_happy_sad.shape

(2498, 14246)

## BOW - Logistic Regression

In [380]:
from sklearn.linear_model import LogisticRegression
from sklearn.cross_validation import train_test_split
from sklearn.metrics import accuracy_score 

logreg = LogisticRegression()

X = bag_of_words_happy_sad
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

logreg.fit(X_train, y_train)
logreg_bow_predictions = logreg.predict(X_test)

print(accuracy_score(y_test, logreg_bow_predictions))

0.792


***BOW: LogReg ANALYSIS***
Bag of Words technique logistic regression shows 79% accuracy of whether it predicts a song is "happy" or "sad" based on how BOW evaluated the lyrics.


**BOW => LogReg GridSearchCV in search of best params**

In [365]:
LR_param_grid = {'solver': ['newton-cg', 'sag', 'saga', 'lbfgs'], 'multi_class': ['ovr', 'multinomial']}

In [366]:
logreg_grid = GridSearchCV(logreg, LR_param_grid)

In [367]:
logreg_grid.fit(X_train, y_train)

/Users/Steve/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/sag.py:326: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/Users/Steve/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/sag.py:326: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/Users/Steve/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/sag.py:326: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/Users/Steve/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/sag.py:326: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/Users/Steve/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/sag.py:326: ConvergenceWarning: The max_iter was

GridSearchCV(cv=None, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'solver': ['newton-cg', 'sag', 'saga', 'lbfgs'], 'multi_class': ['ovr', 'multinomial']},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [368]:
logreg_grid.best_params_

{'multi_class': 'ovr', 'solver': 'newton-cg'}

### Optimizing with best params and best features using SelectFromModel

In [369]:
logreg_bow_best = LogisticRegression(multi_class='ovr', solver='newton-cg')

In [372]:
logreg_bow_selector = SelectFromModel(rfc_bow_best)
X_new_logreg = logreg_bow_selector.fit_transform(X, y)

In [373]:
X_new_logreg.shape

(2498, 149)

In [381]:
X = X_new_logreg
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

logreg_bow_best.fit(X_train, y_train)
logreg_bow_best_predictions = logreg_bow_best.predict(X_test)

print('accuracy score: ', accuracy_score(y_test, logreg_bow_best_predictions))
print(confusion_matrix(y_test, logreg_bow_best_predictions))
print(classification_report(list(y_test), list(logreg_bow_best_predictions)))

accuracy score:  0.744
[[166  79]
 [ 49 206]]
             precision    recall  f1-score   support

      happy       0.77      0.68      0.72       245
        sad       0.72      0.81      0.76       255

avg / total       0.75      0.74      0.74       500



### Analysis:
After applying GridSearchCV and SelectFromModel to Bag of Words (SFM chooses 149 words), the accuracy in predicting whether a song is 'happy' or 'sad' decreases from 79% to 74%. 

## BOW - Random Forest

In [294]:
from sklearn.ensemble import RandomForestClassifier

In [353]:
rfc = RandomForestClassifier()

In [354]:
X = bag_of_words_happy_sad
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

rfc.fit(X_train, y_train)
rfc_predictions = rfc.predict(X_test)

print(accuracy_score(y_test, rfc_predictions))

0.76


**BOW-RFC ANALYSIS:** Random Forest on bag of words performed with around a 76% accuracy

### BOW - RFC - GridSearchCV - Getting best parameters

In [355]:
rfc_param_grid = {'n_estimators': [5, 10, 100], 'min_samples_split': [2, 3, 4, 5, 10], 'max_features': ['sqrt', 'log2', 'auto']}

In [356]:
rfc_grid = GridSearchCV(rfc, rfc_param_grid)
rfc_grid.fit(X_train, y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'n_estimators': [5, 10, 100], 'min_samples_split': [2, 3, 4, 5, 10], 'max_features': ['sqrt', 'log2', 'auto']},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [357]:
rfc_grid.best_params_

{'max_features': 'auto', 'min_samples_split': 5, 'n_estimators': 100}

### BOW - RFC - SelectFromModel - finding best features

In [358]:
rfc_bow_best = RandomForestClassifier(n_estimators=100, min_samples_split=5, max_features='auto')

In [360]:
from sklearn.feature_selection import SelectFromModel

rfc_bow_selector = SelectFromModel(rfc_bow_best)
X_new_rfc = rfc_bow_selector.fit_transform(X, y)

In [362]:
X = X_new_rfc
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

rfc_bow_best.fit(X_train, y_train)
rfc_bow_best_predictions = rfc_bow_best.predict(X_test)

print('accuracy score: ', accuracy_score(y_test, rfc_bow_best_predictions))
print(confusion_matrix(y_test, rfc_bow_best_predictions))
print(classification_report(list(y_test), list(rfc_bow_best_predictions)))

accuracy score:  0.802
[[185  64]
 [ 35 216]]
             precision    recall  f1-score   support

      happy       0.84      0.74      0.79       249
        sad       0.77      0.86      0.81       251

avg / total       0.81      0.80      0.80       500



**ANALYSIS:** After applying GridSearchCV and SelectFromModel to find the best parameters and features, Random Forest Classifer increased in accuracy from 76% to around 80-81% in predicting whether a song is 'happy' or 'sad' when Bag of Words is applied to its lyrics. 

In [364]:
#------------------------------------------------------------#

## BOW - Support Vector Machines

In [297]:
from sklearn.svm import SVC

In [298]:
svc = SVC()

In [299]:
X = bag_of_words_happy_sad
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

svc.fit(X_train, y_train)
svc_bow_predictions = svc.predict(X_test)

print(accuracy_score(y_test, svc_bow_predictions))

0.68


** BOW - Support Vector Machines ** performed with about a 68% accuracy. 

### BOW- SVC - GridSearchCV - finding best params

In [316]:
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [0.1,1, 10], 'gamma': [1,0.1,0.01,0.001]}

In [317]:
grid = GridSearchCV(svc, param_grid)
grid.fit(X_train, y_train)

GridSearchCV(cv=None, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.1, 1, 10], 'gamma': [1, 0.1, 0.01, 0.001]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [319]:
grid.best_params_

{'C': 1, 'gamma': 1}

### BOW- SVC - Select K Best - optimizing

In [345]:
from sklearn.feature_selection import SelectKBest
selector = SelectKBest(k=100)
X_new = selector.fit_transform(X, y)

In [346]:
X_new.shape

(2498, 100)

In [347]:
X = X_new
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [348]:
from sklearn.metrics import confusion_matrix, classification_report

svc_bow_optimized = SVC(C=1, gamma=1)
svc_bow_optimized.fit(X_train, y_train)
svc_bow_op_predictions = svc_bow_optimized.predict(X_test)
print(confusion_matrix(y_test, svc_bow_op_predictions))
print(classification_report(list(y_test), list(svc_bow_op_predictions)))

[[175  68]
 [ 64 193]]
             precision    recall  f1-score   support

      happy       0.73      0.72      0.73       243
        sad       0.74      0.75      0.75       257

avg / total       0.74      0.74      0.74       500



### Optimizing SVM Analysis
After finding the best params with GridSearchCV {'C': 1, 'gamma': 1} and the best 100 words with SelectKBest and thereby optimizing the Support Vector Classifier, its accuracy in predicting whether a song was 'happy' or 'sad' with Bag of Words increased from 68% to 74%. 

## BOW: K-Nearest Neighbors

In [300]:
from sklearn.neighbors import KNeighborsClassifier

In [302]:
knn = KNeighborsClassifier(n_neighbors=3)

In [349]:
X = bag_of_words_happy_sad
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

knn.fit(X_train, y_train)
knn_bow_predictions = knn.predict(X_test)

print(accuracy_score(y_test, knn_bow_predictions))

0.548


__________________________________________________________________
***——————————***////////////\\\\\\\\\\\\\\||||||||||||||||||||||||||||||||||//////////////\\\\\\\\\\\\\\***——————————***
__________________________________________________________________

# TFIDF

In [307]:
tf_idf_vectorizer = TfidfVectorizer()
tf_idf = tf_idf_vectorizer.fit_transform(songs_df['cleaned_lyrics'])

In [1]:
#print(tf_idf)

## Logistic Regression

In [309]:
## performing tfidf on the dataset with happy or sad mood
tf_idf_happy_sad = tf_idf_vectorizer.fit_transform(happy_or_sad_DF['cleaned_lyrics'])

In [311]:
X = tf_idf_happy_sad
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

logreg.fit(X_train, y_train)
logreg_tfidf_predictions = logreg.predict(X_test)

print(accuracy_score(y_test, logreg_tfidf_predictions))

0.76


**TFIDF Logistic Regression:** Logistic regression model on the "TFidf" tool is able to predict whether a song is 'happy' or 'sad' with around 76% accuracy.

## Random Forest

In [217]:
X = tf_idf_happy_sad
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

rfc.fit(X_train, y_train)
rfc_tfidf_predictions = rfc.predict(X_test)

print(accuracy_score(y_test, rfc_tfidf_predictions))

0.72


**Random Forest on TFIDF:** able to predict whether 'happy' or 'sad' with a 72% accuracy. 

## Support Vector Machines

In [312]:
X = tf_idf_happy_sad
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

svc.fit(X_train, y_train)
svc_tfidf_predictions = svc.predict(X_test)

print(accuracy_score(y_test, svc_tfidf_predictions))

0.472


**SVM on TFIDF:** able to predict whether 'happy' or 'sad' with a 47.2% accuracy. 

## K-Nearest Neighbors

In [225]:
X = tf_idf_happy_sad
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

knn.fit(X_train, y_train)
tf_idf_predictions = knn.predict(X_test)

print(accuracy_score(y_test, tf_idf_predictions))

0.628


**K-Nearest Neighbors on TFIDF:** predicts whether a song is 'happy' or 'sad' based on its lyrics with with around a 63% accuracy.  

__________________________________________________________________
***——————————***////////////\\\\\\\\\\\\\\||||||||||||||||||||||||||||||||||//////////////\\\\\\\\\\\\\\***——————————***
__________________________________________________________________

# Doc2Vec

In [226]:
from gensim.models import Doc2Vec
%run Doc2VecHelperFunctions.ipynb

In [2]:
#songs_df.head(20)

In [236]:
convert_lyrics_to_d2v(songs_df['cleaned_lyrics'])

In [237]:
model = Doc2Vec.load('./song_lyrics.d2v')

In [238]:
model.wv.most_similar('good')

[('bad', 0.5962343215942383),
 ('damn', 0.5068840980529785),
 ('fine', 0.45903322100639343),
 ('real', 0.45743805170059204),
 ('nice', 0.4509486258029938),
 ('thank', 0.4380914568901062),
 ('lot', 0.4204558730125427),
 ('better', 0.4187021255493164),
 ('pretti', 0.4120093584060669),
 ('right', 0.4102294445037842)]

In [239]:
model.wv.most_similar('hot')

[('cold', 0.5237383842468262),
 ('sweat', 0.49747610092163086),
 ('spot', 0.4892292618751526),
 ('twineun', 0.4588363468647003),
 ('burnin', 0.4342363476753235),
 ('cock', 0.434071809053421),
 ('ice', 0.43024706840515137),
 ('chilli', 0.4253436028957367),
 ('jumpin', 0.4180157482624054),
 ('block', 0.41756197810173035)]

In [240]:
#songs_df.drop(labels=['_id', 'decades', 'picture'], axis=1, inplace=True)

In [242]:
songs_df.head()

level_0  index                                   _id  \
0        0      0  {'$oid': '52fdfb440b9398049f3d7a8c'}   
1        1      1  {'$oid': '52fdfb3d0b9398049f3cbc8e'}   
2        2     10  {'$oid': '52fdfb430b9398049f3d52a7'}   
3        3    100  {'$oid': '52fdfb3e0b9398049f3ccb47'}   
4        4   1000  {'$oid': '52fdfb400b9398049f3d0bf1'}   

                          album       artist  \
0         Gangnam Style (강남스타일)          PSY   
1                        Native  OneRepublic   
2                     Na Balada  Michel Teló   
3  Good Girl Gone Bad: Reloaded      Rihanna   
4                  Lost In Love   Air Supply   

                                      audio_features  \
0  [11, 0.912744, 0.083704, 132.069, 0.293137, 0....   
1  [6, 0.7457039999999999, 0.11995499999999999, 1...   
2  [11, 0.9030549999999999, 0.6293489999999999, 9...   
3  [9, 0.46322199999999997, 0.09795799999999999, ...   
4  [0, 0.603403, 0.969231, 100.292, 0.027052, 0.3...   

                                      cleaned_lyrics             context  \
0  oppa gangnam style gangnam style najeneun ttas...          [work out]   
1  late ve ve lose sleep dream thing babi ve ve p...         [energetic]   
2  nossa nossa assim você mata ai se eu te pego a...             [party]   
3  ohh bout round applaus yeah standin ovat oh ye...             [party]   
4  m lie head phone think till hurt know hurt tor...  [sexy, sexy, sexy]   

   decades   genres                                    lyrics_features  \
0       []    [pop]  [oppa, gangnam, style, gangnam, style, najeneu...   
1   [2012]    [pop]  [lately, i, ve, been, i, ve, been, losing, sle...   
2       []  [latin]  [nossa, nossa, assim, você, me, mata, ai, se, ...   
3       []    [r&b]  [ohh, how, bout, a, round, of, applause, yeah,...   
4  [1980s]    [pop]  [i, m, lying, alone, with, my, head, on, the, ...   

                       moods                   name       new_context  \
0  [energetic, motivational]  Gangnam Style (강남스타일)          work out   
1                    [happy]         Counting Stars         energetic   
2                [seductive]       Ai Se Eu Te Pego             party   
3       [happy, celebratory]             Take A Bow             party   
4                     [lush]        All Out Of Love  candlelit dinner   

                                             picture  recording_id  \
0  http://images.musicnet.com/albums/073/463/405/...       50232.0   
1  http://images.musicnet.com/albums/081/851/887/...        5839.0   
2  http://images.musicnet.com/albums/066/799/661/...       66836.0   
3  http://images.musicnet.com/albums/057/717/245/...       39446.0   
4  http://images.musicnet.com/albums/068/565/045/...       20005.0   

                                    sub_context        yt_id    yt_views  
0                         [working out: cardio]  9bZkp7q19f0  2450112089  
1                                [energy boost]  hT_nvWreIhg  1020297206  
2                          [dance party: beach]  hcm55lU9knw   657798989  
3                             [girls night out]  J3UjJ4wKLkg   196775394  
4  [making out, slow dancing, romantic evening]  JWdZEumNRmI    24105692

In [243]:
simplified_moods2 = []
for mood in songs_df['moods']:
    if 'happy' in mood:
        simplified_moods2.append('happy')
    elif 'sad' in mood:
        simplified_moods2.append('sad')
    else: 
        simplified_moods2.append(mood)
len(simplified_moods2)

20931

In [245]:
se3 = pd.Series(simplified_moods2)
songs_df['simplified_moods'] = se3.values

In [248]:
songs_df.head()

level_0  index                                   _id  \
0        0      0  {'$oid': '52fdfb440b9398049f3d7a8c'}   
1        1      1  {'$oid': '52fdfb3d0b9398049f3cbc8e'}   
2        2     10  {'$oid': '52fdfb430b9398049f3d52a7'}   
3        3    100  {'$oid': '52fdfb3e0b9398049f3ccb47'}   
4        4   1000  {'$oid': '52fdfb400b9398049f3d0bf1'}   

                          album       artist  \
0         Gangnam Style (강남스타일)          PSY   
1                        Native  OneRepublic   
2                     Na Balada  Michel Teló   
3  Good Girl Gone Bad: Reloaded      Rihanna   
4                  Lost In Love   Air Supply   

                                      audio_features  \
0  [11, 0.912744, 0.083704, 132.069, 0.293137, 0....   
1  [6, 0.7457039999999999, 0.11995499999999999, 1...   
2  [11, 0.9030549999999999, 0.6293489999999999, 9...   
3  [9, 0.46322199999999997, 0.09795799999999999, ...   
4  [0, 0.603403, 0.969231, 100.292, 0.027052, 0.3...   

                                      cleaned_lyrics             context  \
0  oppa gangnam style gangnam style najeneun ttas...          [work out]   
1  late ve ve lose sleep dream thing babi ve ve p...         [energetic]   
2  nossa nossa assim você mata ai se eu te pego a...             [party]   
3  ohh bout round applaus yeah standin ovat oh ye...             [party]   
4  m lie head phone think till hurt know hurt tor...  [sexy, sexy, sexy]   

   decades   genres                                    lyrics_features  \
0       []    [pop]  [oppa, gangnam, style, gangnam, style, najeneu...   
1   [2012]    [pop]  [lately, i, ve, been, i, ve, been, losing, sle...   
2       []  [latin]  [nossa, nossa, assim, você, me, mata, ai, se, ...   
3       []    [r&b]  [ohh, how, bout, a, round, of, applause, yeah,...   
4  [1980s]    [pop]  [i, m, lying, alone, with, my, head, on, the, ...   

                       moods                   name       new_context  \
0  [energetic, motivational]  Gangnam Style (강남스타일)          work out   
1                    [happy]         Counting Stars         energetic   
2                [seductive]       Ai Se Eu Te Pego             party   
3       [happy, celebratory]             Take A Bow             party   
4                     [lush]        All Out Of Love  candlelit dinner   

                                             picture  recording_id  \
0  http://images.musicnet.com/albums/073/463/405/...       50232.0   
1  http://images.musicnet.com/albums/081/851/887/...        5839.0   
2  http://images.musicnet.com/albums/066/799/661/...       66836.0   
3  http://images.musicnet.com/albums/057/717/245/...       39446.0   
4  http://images.musicnet.com/albums/068/565/045/...       20005.0   

                                    sub_context        yt_id    yt_views  \
0                         [working out: cardio]  9bZkp7q19f0  2450112089   
1                                [energy boost]  hT_nvWreIhg  1020297206   
2                          [dance party: beach]  hcm55lU9knw   657798989   
3                             [girls night out]  J3UjJ4wKLkg   196775394   
4  [making out, slow dancing, romantic evening]  JWdZEumNRmI    24105692   

            simplified_moods  
0  [energetic, motivational]  
1                      happy  
2                [seductive]  
3                      happy  
4                     [lush]

In [249]:
#making X
songs_df_happy = songs_df[(songs_df['simplified_moods'] == 'happy')]
songs_df_sad = songs_df[(songs_df['simplified_moods'] == 'sad')]


In [271]:
songs_df_happy_sad = pd.concat((songs_df_happy.sample(n=1249, replace=True), songs_df_sad))

In [4]:
#list(songs_df_happy_sad.index.values)

In [313]:
#Extracting the 'happy' or 'sad' mood songs from the Doc2Vec model for the X values
X_model = []
for index in list(songs_df_happy_sad.index.values):
    X_model.append(model[index])

In [3]:
#X_model

In [266]:
X = X_model
y = songs_df_happy_sad['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

logreg.fit(X_train, y_train)
logreg_d2v_predictions = logreg.predict(X_test)

print(accuracy_score(y_test, logreg_d2v_predictions))

0.724


## Analysis: Happy-Sad Doc2Vec Logistic Regression:
had a 72% accuracy with the doc2vec technique of predicting whether a song was happy or sad based on the vectorization of a happy/sad song's lyrics. 

## Doc2Vec SVM

In [268]:
X = X_model
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

svc.fit(X_train, y_train)
svc_d2v_predictions = svc.predict(X_test)

print(accuracy_score(y_test, svc_d2v_predictions))

0.78


## Doc2Vec Random Forest

In [269]:
X = X_model
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

rfc.fit(X_train, y_train)
rfc_d2v_predictions = rfc.predict(X_test)

print(accuracy_score(y_test, rfc_d2v_predictions))

0.76


## Doc2Vec K-Nearest Neighbors

In [270]:
X = X_model
y = happy_or_sad_DF['simplified_moods']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

knn.fit(X_train, y_train)
knn_d2v_predictions = knn.predict(X_test)

print(accuracy_score(y_test, knn_d2v_predictions))

0.684


## Analysis: 
**Predicting Whether "Happy" or Sad" Song for Doc2Vec:** SVM had the highest accuracy score with 78%, with Random Forest coming in at 76%, logistic regression at 72%, and finally K-Nearest Neighbors at 68%.